rfcs/0006-git-repo-passwordless-auth/README.md

```diff
@@ -0,0 +1,214 @@
+      "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+      "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/<service-account-id>%40my-google-project.iam.gserviceaccount.com"
+    }
+```
+
+### GitHub
+
+Git repositories hosted on GitHub can be accessed via [GitHub Apps](https://docs.github.com/en/apps/overview).
+This allows users to create a single resource from which they can access all
+their GitHub repositories. The app must have sufficient permissions to be able
+to access repositories. The app's ID, private key and installation ID should
+be mentioned in the Secret referred to by `.spec.secretRef`. GitHub Enterprise
+users will also need to mention their GitHub API URL in the Secret.
+
+```yaml
+apiVersion: source.toolkit.fluxcd.io/v1
+kind: GitRepository
+metadata:
+  name: github-repo
+spec:
+  interval: 1m
+  url: https://github.com/<org>/<repository>
+  ref:
+    branch: master
+  provider: github
+  secretRef:
+    name: github-app
+---
+kind: Secret
+metadata:
+  name: gcp-sa
+stringData:
+  githubAppID: <app-id>
+  githubInstallationID: <installation-id>
+  githubPrivateKey: |
+    <PEM-private-key>
+  githubApiURl: <github-enterprise-api-url> #optional, required only for GitHub Enterprise users
+```
+
+## Design Details
+
+### Azure
+
+If `.spec.provider` is set to `azure`, Flux controllers will use [DefaultAzureCredential](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity#DefaultAzureCredential) to build the workload identity credential. This credential type uses the environment variables injected by the Azure Workload Identity mutating webhook. The [access token from the credential will be then used as a bearer token](https://learn.microsoft.com/en-us/azure/devops/integrate/get-started/authentication/service-principal-managed-identity?view=azure-devops#q-can-i-use-a-service-principal-to-do-git-operations-like-clone-a-repo)
+to perform HTTP bearer authentication.
+
+### GCP
+
+If `.spec.provider` is set to `gcp`, Flux controllers will get the Service
+Account credentials from the specified Secret and use
+[`google.CredentialsFromJSON`](https://pkg.go.dev/golang.org/x/oauth2/google#CredentialsFromJSON)
+to fetch the access token. This access token will be then used as the password
+and the `client_email` as the username to perform HTTP basic authentication.
+
+### GitHub
+
+If `.spec.provider` is set to `github`, Flux controllers will get the app
+details from the specified Secret and use it to [generate an app installation
+token](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-an-installation-access-token-for-a-github-app).
+This token is then used as the password and [`x-access-token` as the username](https://docs.github.com/en/apps/creating-github-apps/registering-a-github-app/choosing-permissions-for-a-github-app#choosing-permissions-for-git-access)
+to perform HTTP basic authentication.
+
```

Comment:
```suggestion

### GitLab

This RFC does not have a proposal for GitLab as GitLab offers [Service accounts](https://docs.gitlab.com/ee/user/profile/service_accounts.html) that already work with Flux. Service accounts are fully user-independent accounts that do not allow accessing the GitLab UI, but behave as user accounts in every other aspect.

The recommendation is to use Flux with a service account, as service accounts can be members of projects and groups and can own deploy keys. If Flux is used with deploy tokens, it's worth noting that deploy tokens are not tied to users.

GitLab Service accounts fully address the identity drawback mentioned in the motivation, and partially address the scalability problem. With respect to scalability, the service account needs to be added as a member of the respective Flux projects or their parent groups, and the service account needs to own and rotate its deploy keys. Disabling the service account invalidates its deploy keys as well. Removing a service account from a project, does not invalidate its deploy keys. Deploy keys owned by a service account can be enabled/disabled separately by project.

```

Reply:
Thanks for this addition. From what I read, users have to create a `service account` and `access token` with the Gitlab API. They also have to handle the rotation of the tokens when an `expiration` is set.  So it's quite different than the supposition here where the cloud provider or an app handle the token management.

Reply:
@souleb  I just figured out the Applications at GitLab seem to work very similarly to Applications at GitHub. The user would need to create an Application either through the UI or [the API](https://docs.gitlab.com/ee/api/applications.html), followingly Flux could use [the OAuth access token for API requests](https://docs.gitlab.com/ee/api/oauth2.html#access-gitlab-api-with-access-token). This still needs to be checked if it would work to retrieve a repo or to access the container registry.

Reply:
@nagyv can the gitlab application be used to generate app installation access token similar to gh - https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-an-installation-access-token-for-a-github-app ?

Reply:
from what I understood, when creating an application there is an attribute for `scopes` and we would have a  response like:

```json
{
    "id":1,
    "application_id": "5832fc6e14300a0d962240a8144466eef4ee93ef0d218477e55f11cf12fc3737",
    "application_name": "MyApplication",
    "secret": "ee1dd64b6adc89cf7e2c23099301ccc2c61b441064e9324d963c46902a85ec34",
    "callback_url": "http://redirect.uri",
    "confidential": true
}
```

The scope that is needed in our case is `scopes=read_repository`. If this scope is set, we should be able to use the `application` in order to retrieve an [access token](https://docs.gitlab.com/ee/api/oauth2.html#authorization-code-flow), which can then be used as defined [here](https://docs.gitlab.com/ee/api/oauth2.html#access-git-over-https-with-access-token) to pull the git repository.

I believe the currently defined api is correct here. As for `github` we would need a secret like:

```yaml
kind: Secret
metadata:
  name: gcp-sa
stringData:
  gitlabAppID: <application_id>
  client_secret: <client_secret>
  redirectURl: <redirect_uri>
```

Reply:
Thank you @nagyv and @souleb. Updated the RFC to include gitlab details as suggested.